src/classes/BDI_DataImportService.cls

```diff
@@ -1725,66 +1725,81 @@ global with sharing class BDI_DataImportService {
-        setStrCampaign.clear();
+        List<Campaign> campaignsToInsert = new List<Campaign>();
+        List<DataImport__c> dataImportsWithNewCampaigns = new List<DataImport__c>();
+        uniqueCampaignNames.clear();
         for (DataImport__c dataImport : listDI) {
             // skip di's that already have an error
             if (dataImport.Status__c == BDI_DataImport_API.bdiFailed) {
                 continue;
             }
 
-            String strName = dataImport.Donation_Campaign_Name__c;
+            String campaignName = dataImport.Donation_Campaign_Name__c;
+            Id campaignId = dataImport.DonationCampaignImported__c;
             Campaign cmp;
-            if (strName != null && setStrCampaign.add(strName)) {
-                if (mapStrNameToCampaignId.get(strName) == null) {
-                    cmp = new Campaign(Name=strName, IsActive=true);
-                    listCmpNew.add(cmp);
-                    listDINew.add(dataImport);
+
+            if (campaignId != null) {
+                // User specified a Campaign lookup in BGE lookup or with an ID on import
+                dataImport.DonationCampaignImportStatus__c = System.Label.bdiMatchedId;
+            } else if (campaignName != null && uniqueCampaignNames.add(campaignName)) {
+                if (campaignNameToCampaignId.get(campaignName) == null) {
+                    if (isDryRun) {
+                        dataImport.DonationCampaignImportStatus__c = System.Label.bdiMatchedNone;
+                    } else {
+                        dataImport.DonationCampaignImportStatus__c = System.Label.bdiCreated;
+                        // Create a new Campaign if no match on Name
+                        cmp = new Campaign(Name = campaignName, IsActive = true);
+                        campaignsToInsert.add(cmp);
+                        dataImportsWithNewCampaigns.add(dataImport);
+                    }
+                } else {
+                    // Park the matched Campaign ID in the lookup only after dry run to get the correct match status
+                    if (!isDryRun) {
+                        dataImport.DonationCampaignImported__c = campaignNameToCampaignId.get(campaignName);
+                    }
+                    dataImport.DonationCampaignImportStatus__c = System.Label.bdiMatched;
                 }
             }
-            if (Test.isRunningTest() && strName != null && strName.startsWith('FailTest')) {
+            if (Test.isRunningTest() && campaignName != null && campaignName.startsWith('FailTest')) {
                 cmp.Name = null;
             }
         }
-        //insert listCmpNew;
-        List<Database.UpsertResult> listUR = UTIL_DMLService.upsertRecords(listCmpNew, false);
-        for (Integer i = 0; i < listUR.size(); i++) {
-            Database.UpsertResult ur = listUR[i];
+
+        List<Database.UpsertResult> upsertResults = UTIL_DMLService.upsertRecords(campaignsToInsert, false);
```

Comment:
When are we updating campaigns? Can this just be an insert?

Reply:
This previously was an insert, and you'll see I removed that commented out line in the diff. I'm not sure why this was changed, but it seemed to be made intentionally. I could dig into the previous diffs to see why this change may have been made and if it could just be an insert.

Reply:
@randi274 I'd still love to know what's going on here. At the very least, it would be good to change the variable name to reflect what's actually happening, but as I don't see a use case for updating campaigns I'm more curious about why we'd be doing an upsert at all.

Reply:
I dug back in the git blame, and it looks like Upsert goes all the way back to the origins of the feature. I know this doesn't explain much, but just providing some historical context. Not sure why it was originally designed to upsert instead of insert. Maybe/likely because other objects are being necessarily upserted and this was copied?